synapse/storage/database.py

```diff
@@ -889,20 +891,24 @@ def simple_upsert_txn_native_upsert(
         txn.execute(sql, list(allvalues.values()))
 
     def simple_upsert_many_txn(
-        self, txn, table, key_names, key_values, value_names, value_values
-    ):
+        self,
+        txn: LoggingTransaction,
+        table: str,
+        key_names: Collection[str],
+        key_values: Collection[Iterable[Any]],
+        value_names: Collection[str],
+        value_values: Iterable[Iterable[str]],
+    ) -> None:
         """
-        Upsert, many times.
+        Upsert, many times, but without native UPSERT support or batching.
```

Comment:
> but without native UPSERT support or batching

I didn't dig through it too much, but the fact that this function does is call `simple_upsert_many_txn_native_upsert` (when the engine supports it), which mentions "using batching where possible", feels like this doc is lying.

Reply:
oops, this was a bad c&P, thanks